editor/tree_search.cpp

```diff
@@ -0,0 +1,492 @@
+	}
+	return p_accum;
+}
+
+// Returns the lower and upper bounds of a substring. Does fuzzy search: Simply looks if words exist in right ordering.
+// Also ignores case if p_search_mask is lowercase. Example:
+// p_searcheable = "TimeLimit 2 sec", p_search_mask = limit 2 sec -> [4,14]. With p_search_mask = "LimiT 2 SEC" or "Limit sec 2" -> [-1,-1]
+TreeSearch::StringSearchIndices TreeSearch::_substring_bounds(const String &p_searchable, const String &p_search_mask) const {
+	StringSearchIndices result;
+	result.lower = UPPER_BOUND;
+	result.upper = 0;
+
+	if (p_search_mask.is_empty()) {
+		return result; // Early return if search_mask is empty.
+	}
+
+	// Determine if the search should be case-insensitive.
+	bool is_case_insensitive = (p_search_mask == p_search_mask.to_lower());
+	String searchable_processed = is_case_insensitive ? p_searchable.to_lower() : p_searchable;
+	PackedStringArray words = p_search_mask.split(" ");
+	int word_position = 0;
+	for (const String &word : words) {
+		if (word.is_empty()) {
+			continue; // Skip empty words.
+		}
+
+		String word_processed = is_case_insensitive ? word.to_lower() : word;
+
+		// Find the position of the next word in the searchable string.
+		word_position = searchable_processed.find(word_processed, word_position);
+
+		if (word_position < 0) {
+			// If any word is not found, return an empty StringSearchIndices.
+			return StringSearchIndices();
+		}
+
+		// Update lower and upper bounds.
+		result.lower = MIN(result.lower, word_position);
+		result.upper = MAX(result.upper, static_cast<int>(word_position + word.length()));
+	}
+
+	return result;
+}
+
+void TreeSearch::_select_item(TreeItem *p_item) {
+	if (!p_item)
+		return;
+	ERR_FAIL_COND(!tree_reference || p_item->get_tree() != tree_reference);
+
+	// first unfold ancestors
+	TreeItem *ancestor = p_item->get_parent();
+	while (ancestor) {
+		ancestor->set_collapsed(false);
+		ancestor = ancestor->get_parent();
+	}
+	//then scroll to [item]
+	tree_reference->scroll_to_item(p_item);
+
+	// ...and select it
+	tree_reference->deselect_all();
+	tree_reference->set_selected(p_item, 0);
+}
+
+void TreeSearch::_select_first_match() {
+	if (matching_entries.size() == 0) {
+		return;
+	}
+	for (int i = 0; i < ordered_tree_items.size(); i++) {
+		TreeItem *item = ordered_tree_items[i];
+		if (!_vector_has_bsearch(matching_entries, item)) {
+			continue;
+		}
+		String debug_string = "[";
+		_select_item(item);
+		return;
+	}
+}
+
+void TreeSearch::_select_next_match() {
+	if (matching_entries.size() == 0) {
+		return;
+	}
+	TreeItem *selected = tree_reference->get_selected(); // we care about a single item here.
+	if (!selected) {
+		_select_first_match();
+		return;
+	}
+
+	int selected_idx = -1;
+	for (int i = 0; i < ordered_tree_items.size(); i++) {
+		if (ordered_tree_items[i] == selected) {
+			selected_idx = i;
+			break;
+		}
+	}
+
+	// find the best fitting entry.
+	for (int i = 0; i < ordered_tree_items.size(); i++) {
```

Comment:
Suggestion: Start a loop from `MAX(0, selected_idx)` and drop the second part of the condition below

Reply:
new version:
```cpp
// find the best fitting entry.
for (int i = MAX(0, selected_idx) + 1; i < ordered_tree_items.size(); i++) {
  TreeItem *item = ordered_tree_items[i];
  if (_vector_has_bsearch(matching_entries, item)) {
	  _select_item(item);
	  return;
}
}
```